src/hooks/useCollaborationState.js

```diff
@@ -0,0 +1,92 @@
+import { useState, useEffect, useRef } from 'react'
+
+let iterations = 0
+
+/**
+ * The complete Triforce, or one or more components of the Triforce.
+ * @typedef {CollabState}
+ * @property {boolean} isDirty - Is there something local still waiting to be sync?
+ * @property {Date} dirtySinc - Date of the older object waiting to be sync
+ * @property {Date} lastSave - Date of the earlier object that has been correctly sync
+ */
+
+/**
+ * Monitor the state of collaboration (React hook)
+ *
+ * Returns if something is still in process,
+ * from which time, and the time of the last
+ * successful save to the collab server
+ *
+ * @param {CollabProvider} provider
+ * @returns {CollabState}
+ */
+function useCollaborationState(provider) {
+  const channel = provider && provider.channel
+
+  // using a ref and not a state because often the event
+  // are triggering on and off before the next rendering
+  // in  React, and that introduce an incoherent state
+  // (as each event should use the previous state to
+  // determine its behaviour)
+  const dirtySince = useRef(null)
+  const lastSave = useRef(null)
+  // the state is only here to trigger a rerender when needed
+  const [, setChanges] = useState(iterations)
+
+  // init
+  if (channel) {
+    if (channel.isDirty()) {
+      // always mark as dirty if the channel is dirty
+      if (!dirtySince.current) dirtySince.current = new Date()
```

Comment:
Can `dirtySince.current` even be set at this stage?

Reply:
Most probably, you always start with a non dirty channel. I prefer to be defensive a prepare the case where the channel is dirty the first time I am executing this hook.

However, this specific line may well be executed even after initialization. If proseMirror enqueues a step, we'll fire an event. This event has to be catched. The event listener will then change de ref value and change the state to call a rerender.

All this is async. We may well have a rerender of the same component at any time, before all this, after all this, or between any step in the chain. So yes, we could have a dirty channel but still not have changed the ref value